wallet/src/store/states.rs

```diff
@@ -0,0 +1,376 @@
+    pub fn iter(&self) -> StateIter<'_, K, S> {
+        StateIter {
+            forward: self.head,
+            backward: self.tail,
+            len: self.len(),
+            _anchor: std::marker::PhantomData,
+        }
+    }
+
+    /// access the confirmed state of the store verse
+    pub fn confirmed_state(&self) -> (&K, &S) {
+        if let Some(state) = unsafe { self.head.as_ref() } {
+            debug_assert!(state.confirmed());
+            (&state.key, &state.state)
+        } else {
+            unsafe { std::hint::unreachable_unchecked() }
+        }
+    }
+
+    /// get the last state of the store
+    pub fn last_state(&self) -> (&K, &S, Status) {
+        let key = unsafe { &(*self.tail).key as &K };
+        let state = unsafe { &(*self.tail).state as &S };
+        let status = unsafe { (*self.tail).status };
+
+        (key, state, status)
+    }
+}
+
+impl<K: Hash> Hash for KeyRef<K> {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        unsafe { (*self.0).hash(state) }
+    }
+}
+
+impl<K: PartialEq> PartialEq for KeyRef<K> {
+    fn eq(&self, other: &KeyRef<K>) -> bool {
+        unsafe { (*self.0).eq(&*other.0) }
+    }
+}
+
+impl<K: Eq> Eq for KeyRef<K> {}
+
+#[cfg(not(feature = "nightly"))]
+impl<K> Borrow<K> for KeyRef<K> {
+    fn borrow(&self) -> &K {
+        unsafe { &*self.0 }
+    }
+}
+
+impl<'a, K, S> IntoIterator for &'a States<K, S> {
+    type Item = (&'a K, &'a S, Status);
+    type IntoIter = StateIter<'a, K, S>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.iter()
+    }
+}
+
+impl<'a, K, S> Iterator for StateIter<'a, K, S> {
+    type Item = (&'a K, &'a S, Status);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.len == 0 {
+            return None;
+        }
+
+        let key = unsafe { &(*self.forward).key as &K };
+        let state = unsafe { &(*self.forward).state as &S };
+        let status = unsafe { (*self.forward).status };
+
+        self.len -= 1;
+        self.forward = unsafe { (*self.forward).next };
+
+        Some((key, state, status))
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        (self.len, Some(self.len))
+    }
+
+    fn count(self) -> usize {
+        self.len
+    }
+}
+
+impl<'a, K, S> DoubleEndedIterator for StateIter<'a, K, S> {
+    fn next_back(&mut self) -> Option<Self::Item> {
+        if self.len == 0 {
+            return None;
+        }
+
+        let key = unsafe { &(*self.backward).key as &K };
+        let state = unsafe { &(*self.backward).state as &S };
+        let status = unsafe { (*self.backward).status };
+
+        self.len -= 1;
+        self.backward = unsafe { (*self.backward).prev };
+
+        Some((key, state, status))
+    }
+}
+
+impl<'a, K, S> FusedIterator for StateIter<'a, K, S> {}
+impl<'a, K, S> ExactSizeIterator for StateIter<'a, K, S> {}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn forward_iterator() {
+        let mut multiverse = States::new(0u8, ());
```

Comment:
this is actually more like a `universe` than a `multiverse`, I think. There is just one temporal line.

Reply:
For now in the state. however it is not impossible to extend this to maintain a multiverse of State. The UTxO store allows for this already. There's little to do in order to make the multiverse a reality.